Allow providing of workspace

davidcollom opened the pull request:
This closes #3 Allowing for a configuration or runtime flag to provide the name of the workspace you wish to export from.

I've added a workspace select when logging in..

https://asciinema.org/a/FiezcmTvqvBbDK1oX3BJYLuG2

and a new `clockeditup select` subcommand to switch between workspaces.

maelvls commented on main.go:
```diff
@@ -239,18 +246,26 @@ func Run(tokenFlag string, printHelp func(bool) func()) error {
+		return fmt.Errorf("no workspaces found")
+	}
+
+	workspaceName := workspaceFlag
+	if workspaceName == "" {
+		workspaceName = conf.Workspace
 	}
```
suggestion: I like that in most CLIs, passing a flag "overrides" any env var, and any env var overrides the configuration file. What do you think?

In this context, that would mean inverting the order:

```suggestion
	workspaceName := conf.Workspace
	if workspaceName == "" {
		workspaceName = workspaceFlag
	}
```

maelvls commented on main.go:
```diff
@@ -162,6 +163,12 @@ func Run(tokenFlag string, printHelp func(bool) func()) error {
 
+		conf, err = askWorkspace(conf)
+		if err != nil {
+			return fmt.Errorf("Unable to set workspace: %s", err)
```
nit: 
```suggestion
			return fmt.Errorf("unable to set workspace: %s", err)
```

maelvls commented on clockify.go:
```diff
@@ -135,6 +135,21 @@ func (c *Clockify) Workspaces() ([]Workspace, error) {
 	return workspaces, nil
 }
 
+func (c *Clockify) FindWorkspace(workspaces []Workspace, name string) (Workspace, error) {
```
Looks like this function can stand alone:
```suggestion
func findWorkspace(workspaces []Workspace, name string) (Workspace, error) {
```

I would also suggest to use a `bool` instead of an `error`:
```suggestion
func findWorkspace(workspaces []Workspace, name string) (Workspace, bool) {
```

maelvls commented:
@davidcollom I think that the behavior I implemented ("pick the first workspace") is totally wrong and it is worth changing now. What do you think about this:

If the `workspace` field is not present in `~/.config/clockidup.yaml`, we say to the user to either run `clockidup login`, or use `--workspace`, or use `clockidup select` which will prompt a list of workspaces that can be used (and then save to `~/.config/clockidup.yaml`)

Let's break the current behavior 😅

davidcollom commented:
@maelvls I've implemented the `clockeditup select` subcommand, and added some additional logging/errors if a workspace is not selected.

`clockeditup login` will now ask to select a workspace also and the `-workspace` flag will override any config values.

davidcollom commented on config.go:
```diff
@@ -13,7 +13,7 @@ func saveConfig(pathRelativeToHome string, c Config) error {
 	}
 
-	f, err := os.OpenFile(home+"/"+pathRelativeToHome, os.O_WRONLY|os.O_CREATE, 0600)
+	f, err := os.OpenFile(home+"/"+pathRelativeToHome, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
```
This was causing issues if we had smaller/shorter tokens or workspace names, ending up with invalid yaml.

maelvls replied:
Not sure what caused the invalid yaml and how `os.O_TRUNC` resolved the issue 😅 Can you expand on that?

davidcollom replied:
Yea, so if I have a workspace of 20 characters.. and I switched workspace to one that had 13-14, because we were never truncating the file, the remainder where put at the end of the file after a new line/carriage return.

Under certain conditions (i.e if the remainder of the string contained a space, would become invalid and throw a Marshall failure of the file on the following and all subsequent commands

maelvls replied:
I was able to reproduce the issue after logging in:

```
% clockidup login
error: could not load config: decoding '/home/mvalais/.config/clockidup.yml' from YAML: yaml: line 4: could not find expected ':'
```

As you explained, `~/.config/clockidup.yml` is broken:

```
token: redacted
workspace: aasdasd
-mael-valais
```

What bugs me is that `os.OpenFile` gives a file descriptor where the write pointer is set to the start of the file (i.e., ~~it uses `O_TRUNC` by default~~). I reproduced this behavior [on the Go Playground](https://play.golang.org/p/QZizSlMuCcn).

maelvls replied:
Aaaahhh I forgot that `O_TRUNC` not only sets the write pointer to the start of the file but also _truncates_ the file, i.e., it sets the byte length to 0.

Before writing the configuration, the file was:

```
token: redacted
workspace: jetstack-mael-valais
```

Without `O_TRUNC`, the length of the file stays the same, which means there are some leftovers from the previous file:

```
token: redacted
workspace: aasdasd
-mael-valais
```

With `O_TRUNC`, the leftover disappears:


```
token: redacted
workspace: aasdasd
```

maelvls replied:
**suggestion:** let's add a comment about that:

```suggestion
	// O_TRUNC allows us to re-write the whole configuration file every time we
	// save it. Without it, the file would never shrink in size when writing a
	// shorter configuration manifest, and the YAML manifest would break.
	f, err := os.OpenFile(home+"/"+pathRelativeToHome, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
```

davidcollom reviewed (commented):
Looks good to me! 👍 , If you're happy, I am too! 🎉

maelvls reviewed (approved):
Lgtm!! Thank you!!
